pkg/helm/helm.go

```diff
@@ -18,165 +18,163 @@ package helm
 
 import (
 	"fmt"
-	"io"
-	"io/ioutil"
-	"net/http"
-	"os"
-	"path/filepath"
-
-	"github.com/pkg/errors"
-	"k8s.io/helm/pkg/chartutil"
-	"k8s.io/helm/pkg/getter"
-	"k8s.io/helm/pkg/helm/environment"
-	"k8s.io/helm/pkg/helm/helmpath"
-	"k8s.io/helm/pkg/proto/hapi/chart"
-	"k8s.io/helm/pkg/repo"
-	klog "k8s.io/klog/v2"
+
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/getter"
+	"helm.sh/helm/v3/pkg/helmpath"
+	"helm.sh/helm/v3/pkg/repo"
+
+	"github.com/kubernetes-sigs/minibroker/pkg/log"
 )
 
-const stableURL = "https://kubernetes-charts.storage.googleapis.com"
+const (
+	stableURL = "https://kubernetes-charts.storage.googleapis.com"
+	// As old versions of Kubernetes had a limit on names of 63 characters, Helm uses 53, reserving
+	// 10 characters for charts to add data.
+	helmMaxNameLength = 53
+)
 
+// Client represents a Helm client to interact with the k8s cluster.
 type Client struct {
-	repoURL string
-	home    helmpath.Home
-	rf      *repo.RepoFile
+	log              log.Verboser
+	repositoryClient RepositoryInitializeDownloadLoader
+	chartClient      ChartInstallUninstaller
+
+	settings  *cli.EnvSettings
+	chartRepo *repo.ChartRepository
 }
 
-func NewClient(repoURL string) *Client {
-	if repoURL == "" {
-		repoURL = stableURL
-	}
+// NewDefaultClient creates a new Client with the default dependencies.
+func NewDefaultClient() *Client {
+	return NewClient(
+		log.NewKlog(),
+		NewDefaultRepositoryClient(),
+		NewDefaultChartClient(),
+	)
+}
 
-	return &Client{repoURL: repoURL}
+// NewClient creates a new client with explicit dependencies.
+func NewClient(
+	log log.Verboser,
+	repositoryClient RepositoryInitializeDownloadLoader,
+	chartClient ChartInstallUninstaller,
+) *Client {
+	settings := &cli.EnvSettings{
+		RegistryConfig:   helmpath.ConfigPath("registry.json"),
+		RepositoryConfig: helmpath.ConfigPath("repositories.yaml"),
+		RepositoryCache:  helmpath.CachePath("repository"),
+	}
+	return &Client{
+		log:              log,
+		repositoryClient: repositoryClient,
+		chartClient:      chartClient,
+		settings:         settings,
+	}
 }
 
-func (c *Client) Init() error {
-	klog.V(5).Infof("helm client: initializing client")
+// Initialize initializes a chart repository.
+// TODO(f0rmiga): be able to handle multiple repositories. How can we handle charts with the same
+// name across repositories?
+// TODO(f0rmiga): add a readiness probe for this initialization process. A health endpoint would be
+// enough.
+func (c *Client) Initialize(repoURL string) error {
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: initializing")
```

Comment:
Yeah, the `c.log.V(x).Get().Log()` call sequence looks kind of bad (especially since none of the call sites seem to really need to delay evaluating the arguments).  Can we use more `klog`-style logging?

Reply:
Fixed with https://github.com/kubernetes-sigs/minibroker/pull/72/commits/3d85bc72b8d45a475884d95503213eb65acbc00a.